src/Lookup/EntityRetrievingClosestReferencedEntityIdLookup.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+namespace Wikibase\DataModel\Services\Lookup;
+
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdValue;
+use Wikibase\DataModel\Entity\PropertyId;
+use Wikibase\DataModel\Services\Entity\EntityPrefetcher;
+use Wikibase\DataModel\Snak\PropertyValueSnak;
+use Wikibase\DataModel\Snak\Snak;
+use Wikibase\DataModel\Statement\StatementListProvider;
+
+/**
+ * Service for getting the closest entity (out of a specified set),
+ * from a given starting entity. The starting entity, and the target entities
+ * are (potentially indirectly, via intermediate entities) linked by statements
+ * with a given property ID, pointing from the starting entity to one of the
+ * target entities.
+ *
+ * @since 3.10
+ *
+ * @license GPL-2.0-or-later
+ * @author Marius Hoch
+ */
+class EntityRetrievingClosestReferencedEntityIdLookup implements ReferencedEntityIdLookup {
+
+	/**
+	 * @var EntityLookup
+	 */
+	private $entityLookup;
+
+	/**
+	 * @var EntityPrefetcher
+	 */
+	private $entityPrefetcher;
+
+	/**
+	 * @var int Maximum search depth: Maximum number of intermediate entities to search through.
+	 */
+	private $maxDepth;
+
+	/**
+	 * @var int Maximum number of entities to retrieve.
+	 */
+	private $maxEntityVisits;
+
+	/**
+	 * Map (entity id => true) of already visited entities.
+	 *
+	 * @var bool[]
+	 */
+	private $alreadyVisited = [];
+
+	/**
+	 * @param EntityLookup $entityLookup
+	 * @param EntityPrefetcher $entityPrefetcher
+	 * @param int $maxDepth Maximum search depth: Maximum number of intermediate entities to search through.
+	 * @param int $maxEntityVisits Maximum number of entities to retrieve during a lookup.
+	 */
+	public function __construct(
+		EntityLookup $entityLookup,
+		EntityPrefetcher $entityPrefetcher,
+		$maxDepth,
+		$maxEntityVisits
+	) {
+		$this->entityLookup = $entityLookup;
+		$this->entityPrefetcher = $entityPrefetcher;
+		$this->maxDepth = $maxDepth;
+		$this->maxEntityVisits = $maxEntityVisits;
+	}
+
+	/**
+	 * Get the closest entity (out of $toIds), from a given entity. The starting entity, and
+	 * the target entities are (potentially indirectly, via intermediate entities) linked by
+	 * statements with the given property ID, pointing from the starting entity to one of the
+	 * target entities.
+	 *
+	 * @since 3.10
+	 *
+	 * @param EntityId $fromId
+	 * @param PropertyId $propertyId
+	 * @param EntityId[] $toIds
+	 *
+	 * @return EntityId|null Returns null in case none of the target entities are referenced.
+	 * @throws ReferencedEntityIdLookupException
+	 */
+	public function getReferencedEntityId( EntityId $fromId, PropertyId $propertyId, array $toIds ) {
+		if ( !$toIds ) {
+			return null;
+		}
+
+		$this->alreadyVisited = [];
```

Comment:
No, because we want to maybe visit the same entities again, if another (or even the same) reference-relationship is queried.
We may have a caching decorator for this in the future, but that's not this classes business.